cosa/20200805-meta-merge.md

```diff
@@ -0,0 +1,103 @@
+
+One of the biggest blockers to performing parallel builds, distributed builds and/or re-starting failed builds is the ability to lock write access to `meta.json` in COSA.
+
+Under the following assmptions, COSA needs to support merging of `meta.json`:
+- Not all `cmd-buildextend-*` operations will happen on the same pod or Linux namespace
+- S3 is the authoritative store for builds. S3 is eventually consistent meaning that the version read from S3 may be replaced after reading from S3. Implementing distributed locking on `meta.json` based on S3 is problematic.
+- For RHCOS we need the ability to do partial builds and then back-fill artifacts (e.g. only build qemu, but not cloud) and later build cloud.
+- Developers want to restart failed builds
+- Consumers of the artifacts from COSA want to reduce the time to build and test
+
+## State of a build is a point-of-view statement
+
+To COSA a build is usually `ostree` + `qcow2`. The notion of a "failed" or "partial" build is a point of view. COSA does provision the build id, but COSA does not know what comprised the artifacts of the "build". Usually developers and consumers speak of a build as the base `ostree` and the _optional_ artifacts. 
+
+In COSA a "build" only starts to exist when the `meta.json` is incepted after the `ostree` has been written. The build directory may exist, but the `meta.json` that is produced immediately is valid for any consumer. In most cases a "failed" build means that an artifact in the `buildextend-*` commands has failed to publish (such as AWS) not that the artifact itself failed to build; the artifact was indeed built. 
+
+Further, a "failed build" is also used to mean that some part of a pipeline failed to complete. COSA itself does not have any concept of dependencies, much less the idea that a multiple COSA commands can be run. COSA only has implied dependencies that some artifacts requires others. In other words, COSA doesn't know when the "build" is complete until after something else considers its complete. 
+
+Denoting the state of a build in the `meta.json` will be owned by process/thing/person that is orchestrating the build. 
+
+## Multiple `meta.json`
+
+Under the axiom that the state of a build is usually determined from the point-of-view of a pipeline and to support distributed and parallel builds, builds will be produced with multiple `meta.json` that will be merged together.
+
+## `cosalib.meta` library
+
+`cmd-meta` currently uses the [`cosalib.meta`](https://github.com/coreos/coreos-assembler/blob/master/src/cosalib/meta.py) library for reading and writing `meta.json`. To support merging of `meta.json` files, the library will be updated to allow for merging in-memory updates with on-disk content.
+
+Only `meta.json` that is validated will be merged under the following rules:
+- the on-disk version must match in core details for name and buildid for in memory model of the meta
+- on-disk meta.json takes precedence over the in-memory meta
```

Comment:
The idea here was that you can't merge FCOS into RHCOS; that why I called out `name` and `buildid`. I wanted to prevent merging 20200101 into 202501010.